core/src/main/java/org/apache/calcite/sql/fun/SqlSubstringFunction.java

```diff
@@ -195,8 +195,9 @@ public void unparse(
       final SqlMonotonicity mono0 = call.getOperandMonotonicity(0);
       if ((mono0 != SqlMonotonicity.NOT_MONOTONIC)
           && call.getOperandMonotonicity(1) == SqlMonotonicity.CONSTANT
-          && call.getOperandLiteralValue(1, BigDecimal.class)
-              .equals(BigDecimal.ZERO)
+          && ((call.isOperandNull(1, true))
+              || call.getOperandLiteralValue(1, BigDecimal.class)
+              .equals(BigDecimal.ZERO))
```

Comment:
I don't think the mono0 is still the same when the operand1 is null, that means `substring('abc', null)` could returns null, which may breaks the monotonicity of operand0.

Reply:
Yes, you are right. I just focused on NPE exception but ignored the semantic.
It seems that we should check null for each SqlOperator to align with the runtime.